Endpoint for Organisations and ContentItems per organisation

pmanrubia opened the pull request:
Covers [this story][1] that exercises the first components of our application. Organisations, Content Items, Controllers and all the testing stack.

## Includes:

1\. Exposes an **endpoint** at `/organisations/index` and return a JSON with the organisations and the number of items of each organisation

```
[{
  organisation_id: 8392378947298374,
  number_of_pages: 12.323
},{
  organisation_id: 8392378947298374,
  number_of_pages: 123
}]
```

2\. Add a **`rake task`** to import all the Content Items for an Organisation provided the Organisation slug

```bash
$ rails import:organisation[hm-revenue-customs]
```

## Notes

1. We temporarily rely on [HTTParty][5] to query SearchAPI; this is a temporary solution as we will use GDS  Adapters as a way to communicate with other applications. The task to upgrade to use GDS adapters is tracked in a [Trello Card][6]
2. It is not ideal the way we are mocking the HTTP requests, as we are not `webmok-ing` the `get` request only the library request. This is something that [will go away in a few weeks][6], so no need to work any further.
3. We are using a very simple mechanism to serialise to JSON the organisations . This is because most of this code will go away once we expose an API to other apps. We will to rely on [Api Serializers][3] or using libraries like [Grape][4]. No need to complicate things further right now.
4. As a result of the development of this PR, I have created a [Trello card][2] to automate Lint when running all tests in each commit so we prevent this situation from happening again. This way we don't need to write a final commit with all the fixes.

## Aditional notes

We are not doing the Heroku work as part of this story because there is some pending work related to use GDS Heroku account and I don't want this task to block this card in our Trello board.
I have created a [Trello card][7] to track the work to deploy to Heroku.

[1]: https://trello.com/c/JP2lwuod/69-as-a-content-designer-i-would-like-to-access-the-number-of-pages-i-that-belong-to-my-organisation
[2]: https://trello.com/c/O6wNOS9N/79-run-lint-validator-when-running-rspec-tests
[3]: https://github.com/rails-api/active_model_serializers
[4]: https://github.com/ruby-grape/grape
[5]: https://github.com/jnunemaker/httparty
[6]: https://trello.com/c/mn8JPHYR/78-use-gds-adapter-to-retrieve-content-items-for-an-organisation
[7]: https://trello.com/c/1Hc4Q5Np/81-deploy-to-the-common-heroku-account-of-gds

leenagupte commented on spec/controllers/organisations_controller_spec.rb:
```diff
@@ -12,8 +12,5 @@
-
   end
-
 end
```
This is an odd place to remove lines in this file.

pmanrubia replied:
Yes, during the live-coding-sessions we didn't put that much attention to spaces and blank lines. This is why I have [fixed all lint errors in the latest commit](https://github.com/alphagov/com-prototype/pull/6/commits/db720509f1e2011fdb17b434536cacffa19d73ce), and I have [create a Trello card](https://trello.com/c/O6wNOS9N/79-run-lint-validator-when-running-rspec-tests) to prevent this situation from happening again.

leenagupte commented on db/schema.rb:
```diff
@@ -0,0 +1,33 @@
+  end
+
+  add_foreign_key "content_items", "organisations"
+end
```
I would move the creation of the schema file to commit 66da862d271577732a621318b33f4165eba73c14

pmanrubia replied:
thanks!, fixed in https://github.com/alphagov/com-prototype/pull/7/commits/58ecd31901d1fb9f8670011f08a287731acb2bd9

leenagupte commented on Gemfile.lock:
```diff
@@ -222,4 +225,4 @@ DEPENDENCIES
 
 BUNDLED WITH
-   1.13.1
+   1.13.6
```
Why has the version changed? Is it a requirement of the shoulda-matchers gem?

pmanrubia replied:
I had a more modern Bundler version. This number never goes backwards.

leenagupte replied:
Ok.

leenagupte commented on Gemfile.lock:
```diff
@@ -230,5 +234,8 @@ DEPENDENCIES
 
+RUBY VERSION
+   ruby 2.3.1p112
+
```
Do we need this? Is this something that should have been added when the app was created?

pmanrubia replied:
I would get rid of it, but it was added in https://github.com/alphagov/com-prototype/pull/2 in https://github.com/alphagov/com-prototype/pull/2/commits/6e0691a054991bad257db1794f86cd15620787a8. 
Not a big deal, though. It might be easier for the first sprints to rely only on `.ruby-version`

leenagupte commented:
In commit aa0254427c508ef0bbd41cf85829754442056cbc can you change the commit message to `filters all organisations by ContentItems`.

`filters out` means to remove the content items.

leenagupte commented on spec/models/importers/organisation_spec.rb:
```diff
@@ -63,4 +63,11 @@
     expect(content_ids).to eq(%w(number-1 number-2))
   end
+
+  it 'raises an exception with an organisation that does not exist' do
```
Should change this to `raises an exception with an organisation that does not have any content items`

leenagupte commented on spec/models/importers/organisation_spec.rb:
```diff
@@ -1,28 +1,19 @@
-          },
-        ]}.to_json)
-      allow(HTTParty).to receive(:get).and_return(response)
+      expect(HTTParty).to receive(:get).and_return(one_content_item_response)
```
Why did this change from `allow` to `expect`?

pmanrubia replied:
We are not mocking the get request, just the library. So, I just wanted to be sure that we are actually using the mock. As commented in the description the library (HTTParty) along with the way we are testing it will go away in a few weeks, so feels good enough for the time being.

pmanrubia replied:
I have updated it. Thank you

leenagupte commented on app/models/importers/organisation.rb:
```diff
@@ -11,19 +12,36 @@ def initialize(slug, batch: 10, start: 0)
-
-      results.each do |result|
-        attributes = result.slice('content_id')
+      result = search_content_items_for_organisation
```
I thought it wasn't very Rubyist to use verbs in method names? Should this be `content_items_for_organisation`?

pmanrubia replied:
My understanding is that `nouns` are good for class names, and `verbs` for methods. The opposite, using verbs for class names is the antipattern.
Have a look at a controller or at any ActiveRecord API. Most of them are verbs.

leenagupte replied:
I shall point @ikennaokpala to this comment :)

leenagupte commented on app/models/importers/organisation.rb:
```diff
@@ -11,19 +12,36 @@ def initialize(slug, batch: 10, start: 0)
+  end
+
+  def next_page!
+    self.start += batch
```
Doesn't `self` mean that you are using the class variable? Isn't `start` an instance variable now? Shouldn't `start` be passed to `next_page!` instead? Or, use `@start`?

leenagupte replied:
Thanks for the explanation. I'm never quite sure when `self` refers to a class variable or the existing instance variable.

leenagupte commented on app/models/importers/organisation.rb:
```diff
@@ -25,7 +24,7 @@ def run
 
-  private
+private
 
```
You could fix this up to the previous commit!

pmanrubia replied:
I just fixed all lint errors in one commit. I prefer not to put time into fixing up this line. I hope this is ok with you.

leenagupte replied:
That's fine.

leenagupte commented on spec/controllers/organisations_controller_spec.rb:
```diff
@@ -1,14 +1,13 @@
     end
 
-    it "returns JSON content type" do
+    it 'returns JSON as content type' do
```
I think this reads better as `returns a JSON content type`

pmanrubia replied:
I have updated the copy of the test. Thank you

leenagupte commented on spec/lib/tasks/import_spec.rb:
```diff
@@ -20,7 +20,7 @@
 
-    it 'raise an error if a slug is not provided' do
+    it 'raise an error if a slug parameter is not present' do
       expect { Rake::Task['import:organisation'].invoke }.to raise_error('Missing slug parameter')
```
Nitpick: should be `raises`.

pmanrubia replied:
I have updated the copy of the test, thank you

leenagupte commented:
@pmanrubia As a whole, the code looks good. I'm missing the story though. 

In commit 5979c3a, an empty `run` method is created to test the new rake task, but this `run` method is ignored for three commits until aa0254427c508ef0bbd41cf85829754442056cbc. 

I thought the plan was to only add something right before it's needed?

pmanrubia commented:
> In commit aa02544 can you change the commit message to filters all organisations by ContentItems. > filters out means to remove the content items.

Done in #7

pmanrubia commented:
Thank you for your review, I have addressed your comments in #7 

Regarding the blocks of commits... I have move 5979c3a down 3 commits to be close to aa02544. I hope this is more clear now.

Closing this PR, as this has been superseeded by #7
